File: redis/cpp/receiver_async_test.cpp
#include <hiredis/hiredis.h>
#include <iostream>
#include <string>
#include <cstring>
#include <signal.h>
#include <thread>
#include <atomic>
#include "../../utils/cpp/message_helpers.hpp"

using messaging::MessageEnvelope;
using message_helpers::get_current_time_ms;

std::atomic<bool> running(true);

void signal_handler(int sig) {
    running = false;
}

int main(int argc, char* argv[]) {
    int receiver_id = 0;
    
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--id") == 0 && i + 1 < argc) {
            receiver_id = std::stoi(argv[i + 1]);
            break;
        }
    }
    
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    
    redisContext *c_sub = redisConnect("127.0.0.1", 6379);
    redisContext *c_pub = redisConnect("127.0.0.1", 6379);
    
    if (c_sub == NULL || c_sub->err || c_pub == NULL || c_pub->err) {
        std::cerr << "Redis connection failed" << std::endl;
        if (c_sub) redisFree(c_sub);
        if (c_pub) redisFree(c_pub);
        return 1;
    }
    std::cout << " [+] [ASYNC] Connected to Redis" << std::endl;

    std::string channel = "test_channel_" + std::to_string(receiver_id);
    std::cout << " [*] [ASYNC] Receiver " << receiver_id << " waiting for messages on " << channel << std::endl;

    // Subscribe to channel
    redisReply *sub = (redisReply*)redisCommand(c_sub, "SUBSCRIBE %s", channel.c_str());
    if (sub) freeReplyObject(sub);

    // Set timeout for graceful shutdown
    struct timeval tv = {1, 0};
    redisSetTimeout(c_sub, tv);

    while (running) {
        redisReply *reply = nullptr;
        int status = redisGetReply(c_sub, (void**)&reply);
        
        if (status == REDIS_OK && reply) {
            if (reply->type == REDIS_REPLY_ARRAY && reply->elements >= 3) {
                if (reply->element[0]->type == REDIS_REPLY_STRING && 
                    strcmp(reply->element[0]->str, "message") == 0) {
                    
                    std::string message_str(reply->element[2]->str, reply->element[2]->len);
                    
                    // Parse message
                    MessageEnvelope msg_envelope;
                    if (message_helpers::parse_envelope(message_str, msg_envelope)) {
                        std::string message_id = msg_envelope.message_id();
                        std::cout << " [x] [ASYNC] Received message " << message_id << std::endl;
                        
                        // Create ACK
                        MessageEnvelope response = message_helpers::create_ack_from_envelope(
                            msg_envelope,
                            std::to_string(receiver_id)
                        );
                        response.set_async(true);
                        std::string response_str = message_helpers::serialize_envelope(response);
                        
                        // Send ACK to reply channel
                        std::string reply_channel = "reply_" + message_id;
                        if (msg_envelope.metadata().count("reply_to")) {
                            reply_channel = msg_envelope.metadata().at("reply_to");
                        }

                        redisReply *pub = (redisReply*)redisCommand(c_pub, "PUBLISH %s %b", 
                            reply_channel.c_str(), response_str.data(), response_str.size());
                        if (pub) freeReplyObject(pub);
                    }
                }
            }
            freeReplyObject(reply);
        } else if (status == REDIS_ERR) {
            // Handle error (including timeout)
            if (c_sub->err == REDIS_ERR_IO && (errno == EAGAIN || errno == EINTR || errno == 0)) {
                c_sub->err = 0;
                memset(c_sub->errstr, 0, sizeof(c_sub->errstr));
            } else if (running) {
                std::cerr << " [!] Redis error: " << c_sub->errstr << " (code: " << c_sub->err << ")" << std::endl;
                std::this_thread::sleep_for(std::chrono::milliseconds(100));
            }
        }
    }

    std::cout << " [x] [ASYNC] Receiver " << receiver_id << " shutting down" << std::endl;
    redisFree(c_sub);
    redisFree(c_pub);
    return 0;
}

File: redis/python/receiver_async_test.py
#!/usr/bin/env python3
"""Redis Python Receiver - Async"""
import sys
import signal
import asyncio
import redis.asyncio as redis
from pathlib import Path

# Add utils to path
repo_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(repo_root / 'utils' / 'python'))

from message_helpers import *

running = True

def signal_handler(sig, frame):
    global running
    running = False

async def run(receiver_id):
    r = redis.Redis(host='localhost', port=6379, db=0)
    pubsub = r.pubsub(ignore_subscribe_messages=True)
    
    channel_name = f"test_channel_{receiver_id}"
    await pubsub.subscribe(channel_name)
    
    print(f" [*] [ASYNC] Receiver {receiver_id} waiting for messages on {channel_name}")
    
    while running:
        try:
            message = await pubsub.get_message(timeout=0.1)
            if message and message['type'] == 'message':
                request_envelope = parse_envelope(message['data'])
                message_id = request_envelope.message_id
                print(f" [x] [ASYNC] Received message {message_id}")
                
                # Create ACK
                response = create_ack_from_envelope(request_envelope, str(receiver_id))
                setattr(response, 'async', True)
                resp_str = serialize_envelope(response)
                
                # Send reply
                if 'reply_to' in request_envelope.metadata:
                    await r.publish(request_envelope.metadata['reply_to'], resp_str)
        except Exception as e:
            print(f"Error: {e}")
            await asyncio.sleep(0.1)
            
    print(f" [x] [ASYNC] Receiver {receiver_id} shutting down")
    await pubsub.close()
    await r.close()


def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--id', type=int, default=0)
    args = parser.parse_args()
    
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
    
    asyncio.run(run(args.id))


if __name__ == "__main__":
    main()
